src/helics/core/FederateState.cpp

```diff
@@ -1671,6 +1671,65 @@ void FederateState::logMessage(int level,
     }
 }
 
+void FederateState::sendCommand(ActionMessage& command)
+{
+    auto cmd = command.payload.to_string();
+    if (cmd == "terminate") {
+        if (parent_ != nullptr) {
+            ActionMessage bye(CMD_DISCONNECT);
+            bye.source_id = global_id.load();
+            bye.dest_id = bye.source_id;
+            parent_->addActionMessage(bye);
+        }
+    } else if (cmd == "echo") {
+        ActionMessage response(CMD_SEND_COMMAND);
+        response.payload = "echo_reply";
+        response.dest_id = command.source_id;
+        response.source_id = global_id.load();
+        response.setString(targetStringLoc, command.getString(sourceStringLoc));
+        response.setString(sourceStringLoc, getIdentifier());
+        if (parent_ != nullptr) {
+            parent_->addActionMessage(response);
+        }
+    } else if (cmd == "command_status") {
+        ActionMessage response(CMD_SEND_COMMAND);
+        response.payload = fmt::format("\"{} unprocessed commands\"", commandQueue.size());
+        response.dest_id = command.source_id;
+        response.source_id = global_id.load();
+        response.setString(targetStringLoc, command.getString(sourceStringLoc));
+        response.setString(sourceStringLoc, getIdentifier());
+        if (parent_ != nullptr) {
+            parent_->addActionMessage(response);
+        }
+    } else {
+        commandQueue.emplace(cmd, command.getString(sourceStringLoc));
+    }
+}
+
+std::pair<std::string, std::string> FederateState::getCommand()
+{
+    auto val = commandQueue.try_pop();
+    if (val->first == "notify") {
+        if (parent_ != nullptr) {
+            parent_->sendCommand(val->second, "notify_response", name);
+        }
+        val = commandQueue.try_pop();
+    }
+    return (val) ? *val : std::pair<std::string, std::string>{std::string{}, std::string{}};
+}
+
+std::pair<std::string, std::string> FederateState::waitCommand()
+{
+    auto val = commandQueue.pop();
+    if (val.first == "notify") {
+        if (parent_ != nullptr) {
+            parent_->sendCommand(val.second, "notify_response", name);
+        }
+        val = commandQueue.pop();
```

Comment:
Why are there multiple calls to pop? Shouldn't the first call already remove the value for the next command from the queue?

Reply:
Notify is a command that shouldn't be sent to the federate,  but generates an echo like response to the source.  The command that gets sent back up to the federate is the next one, hence the two pops in that path.